cpp/harvest/marc_issn_lookup.cc

```diff
@@ -0,0 +1,171 @@
+    for (auto &field : record) {
+        if (field.getTag() == "300") {
+            const std::string sub_field_a(field.getFirstSubfieldWithCode('a'));
+            if (sub_field_a == "Online-Ressource")
+                return true;
+            else
+                return false;
+        }
+    }
+    return false;
+}
+
+SubFieldInfo LookupISSNInDataSouce(const std::string input_journal_filename, const std::string issn) {
+    auto data_source(MARC::Reader::Factory(input_journal_filename));
+    SubFieldInfo sub_field_info = SubFieldInfo(false);
+
+    while (MARC::Record record = data_source->read()) {
+        for (auto &field : record) {
+            if (field.getTag() == "022") {
+                const std::string sub_field_a(field.getFirstSubfieldWithCode('a'));
+                if ((sub_field_a != "") && (sub_field_a == issn)) {
+                    sub_field_info.Builder(record);
+                    // the issn is online version, no need to check further
+                    if (IsOnlineISSN(record))
+                        return sub_field_info;
+                } else {
+                    // maybe the issn is in subfield l
+                    const std::string sub_field_l(field.getFirstSubfieldWithCode('l'));
+                    if ((sub_field_l != "") && (sub_field_l == issn)) {
+                        sub_field_info.Builder(record);
+                        // the issn is online version, no need to check further
+                        if (IsOnlineISSN(record))
+                            return sub_field_info;
+                    }
+                }
+            }
+        }
+    }
+    return sub_field_info;
+}
+
+void UpdateSubfield(MARC::Subfields &subfields, const SubFieldInfo &sub_field_info) {
+    if (!subfields.replaceFirstSubfield('i', sub_field_info.i))
+        subfields.addSubfield('i', sub_field_info.i);
+    if (!subfields.replaceFirstSubfield('x', sub_field_info.x))
+        subfields.addSubfield('x', sub_field_info.x);
+    if (!subfields.replaceFirstSubfield('w', sub_field_info.w))
+        subfields.addSubfield('w', sub_field_info.w);
+
+    if (sub_field_info.t != "") {
+        if (!subfields.replaceFirstSubfield('t', sub_field_info.t))
+            subfields.addSubfield('t', sub_field_info.t);
+    }
+}
+
+void ISSNLookup(char **argv) {
+    auto input_file(MARC::Reader::Factory(argv[1]));
+    auto output_file(MARC::Writer::Factory(argv[3]));
+    int onprogress_counter(1);
+    std::string ppn("");
+
+    std::cout << "Updating in progress...\n";
+    while (MARC::Record record = input_file->read()) {
+        std::cout << "Record - " << onprogress_counter << "\r";
+        ppn = "";
+
+        for (auto &field : record) {
+            if (field.getTag() == "773") {
+                const std::string issn(field.getFirstSubfieldWithCode('x'));
+                if (issn != "") {
+                    SubFieldInfo sub_field_info(LookupISSNInDataSouce(argv[2], issn));
```

Comment:
In our sample data, there are about 3.500 ISSNs (1.500 are unique). If `LookupISSNInDataSouce` is called for every ISSN, this means that the whole `marc_input_journals` file will be read multiple thousand times. I think it would be better to read the file only once (outside the loop) and create some kind of cache, like e.g. a std::map<std::string, SubFieldInfo> where the key is the ISSN to lookup.